### DISCUSSION TITLE
Transferring RtAero Forces and moments to SWL.

### DISCUSSION THREAD

**Discussion by MYMahfouz** [OpenFAST/openfast#1479] [General]
Hi,

I am trying to transfer the Aerodynamic forces and moment from the rotor to the sea water level. I found the discussion [here](https://forums.nrel.gov/t/switching-from-rtaero-to-twrbs-in-an-offshore-test/2365/9?fbclid=IwAR31Q5D3oMKZCz5pFniSKy4APAQAUUHJ4XTI9N0icD4MLeg9zPLzMHgOj0g) and I found it useful.

The main difference is that I am including the rotor uptilt angle , cone angle as well as the overhang. 
My current method at the moment is as follows:
1. First I transfer the aerdynamic forces and moment from the rotating hub reference frame to a fixed reference frame using the following equations.

Fx_nac=RtAeroFxh
Fy_nac=RtAeroFyh * cos(Azimuthpi/180) - RtAeroFzh * sin(Azimuthpi/180)
Fz_nac=RtAeroFyh * sin(Azimuthpi/180) + RtAeroFzh * cos(Azimuthpi/180)
Mx_nac=RotAeroMxh
My_nac=RtAeroMyh * cos(Azimuthpi/180) - RtAeroMzh * sin(Azimuthpi/180)
Mz_nac=RtAeroMzh * cos(Azimuthpi/180)+RtAeroMyh * sin(Azimuthpi/180)

2. Then I rotate the forces around the uptilt angle using the following equations

Fx = Fx_nac * cos(uptilt angle) + Fz_nac * sin(uptilt angle)
Fy = Fy_nac
Fz = Fz_nac * cos(uptilt angle) - Fx_nac * sin(uptilt angle)
Mx = Mx_nac * cos(uptilt angle) + Mz_nac * sin(uptilt angle)
My = My_nac 
Mz = Mz_nac * cos(uptilt angle) - Mx_nac * sin(uptilt angle)

3. Then I multiply the Forces with the overhang arm as follows to get the moments at the tower base:

F_twrbase= [Fx , Fy, Fz]^T
M_twrbase= [Mx, My, Mz]^T + [overhang * cos(uptiltangle), 0, HubHeight]^T x [Fx , Fy, Fz]^T

4. Finally I use a transformation matrix to rotate the tower base coordinate system to a coordinate system parallel to the global coordinate system:

F= TransMat * F_twrBase 
M= TransMat * M_twrBase 

I believe there is something missing already at the second step cause even for a fixed bottom wind turbine my Mz does not match the TwrBsMz in OpenFAST. I believe in the absence of any platform motions these two values should match.
Does this process make sense?
Can you please tell me what is missing in my steps here? Is the uptilt angle alone not enough?

Thanks in advance.

**Comment by jjonkman** [comment]
Dear @MYMahfouz,

I generally follow your approach, except the last step, where, unless TransMat is identify, I would think the rotation would also introduce offsets/moment arms.

Regarding step 1, where are you getting Azimuth from?  Is this the ElastoDyn `Azimuth` output defined relative to `AzimB1Up`?

Regarding step 2, how are defining uptilt?  Your equation looks correct if uptilt is positive for an upwind rotor.  But, typically, shaft tilt is defined negatively for an upwind rotor.

Best regards,

**Comment by MYMahfouz** [comment]
Dear @jjonkman,

Thanks for your reply.

- Step 1: Yes, I am using the Azimuth from ElastoDyn  `Azimuth`.
- Step 2: Yes I am defining the uplift angle as positive and not negative as defined in ElastoDyn.
- This is the TransMat I am using. I am using this matrix because I am using the static dynamic forces as input forces to MoorPy at sea water level. MoorPy uses this matrix for transferring the inertia and bouyancy forces to sea water level.

```python 
# x1: yaw angle
# x2: pitch angle
# x3: roll angle

    s1 = np.sin(x1) 
    c1 = np.cos(x1)
    s2 = np.sin(x2) 
    c2 = np.cos(x2)
    s3 = np.sin(x3) 
    c3 = np.cos(x3)
    
    
    R = np.array([[ c1*c2,  c1*s2*s3-c3*s1,  s1*s3+c1*c3*s2],
                  [ c2*s1,  c1*c3+s1*s2*s3,  c3*s1*s2-c1*s3],
                  [   -s2,           c2*s3,           c2*c3]])
```

I forgot to mention that the results get closer to each other when I include the Tower top SS and FA position to step 3. 
The equation becomes: 
[Mx, My, Mz]^T + [overhang * cos(uptiltangle)+TTDspFA , TTDspSS, HubHeight]^T x [Fx , Fy, Fz]^T

Best regards,

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

A few comments:
- The platform rotations will create additional moment arms because the hub will be in a different location relative to the tower base when the platform rotations are nonzero.
- ElastoDyn uses a different transformation matrix for the platform roll, pitch, and yaw angles as mentioned in the forum post you linked to.
- The tower-top displacements will also induce small rotation of the nacelle relative to the tower base, which you don't seem to be accounting for.
- I would suggest verifying that the transformation works properly without any tower deflection or platform rotation first.

Best regards,

**Comment by MYMahfouz** [comment]
Dear @jjonkman,

Thanks this is really helpful.

My model works well for a rigid tower and fixed platform.


I have changed the structure of my coordinate transformation to follow openfast implementation


I have now implemented the Transformation matrix as you mentioned. I am implementing also the tower top rotations in my code. I have a question:
- I can see in ElastoDyn source code you used `ThetaFA` and `ThetaSS`, are these equivalent to `TTDspPtch` and `TTDspRoll`?


Thanks for the help.

Best regards,

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

Yes, I agree that those are equivalent, with the latter in degrees and the former in radians.

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

Thanks I have one more question for the transformation matrix:
For tower top rotations and platform rotations you are using this

`SmllRotTrans( 'tower deflection (ElastoDyn SetCoordSy)', ThetaSS, 0.0_R8Ki, ThetaFA, TransMat, TRIM(Num2LStr(t))//' s', ErrStat2, ErrMsg2 )`

I see that you are reversing the y axis into the negative y axis. I am using the same input then using the inverse of the matrix. Is this the correct implementation?

Best regards,

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

Yes that is correct.  And because `TransMat` is an orientation matrix/DCM, the matrix inverse is also the matrix transpose.

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,
Thanks.
I have another question for the ThetaFA you used it as +ve in the transformation while the PtfM pitch you used as negative ? I thought both angles should have the same sign, why is this not the case?
`SmllRotTrans( 'tower deflection (ElastoDyn SetCoordSy)', ThetaSS, 0.0_R8Ki, ThetaFA, TransMat, TRIM(Num2LStr(t))//' s', ErrStat2, ErrMsg2 )`

`SmllRotTrans( 'platform displacement (ElastoDyn SetCoordSy)', x%QT(DOF_R), x%QT(DOF_Y), -x%QT(DOF_P), TransMat, TRIM(Num2LStr(t))//' s', ErrStat2, ErrMsg2 )`

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

These are stated correctly.  They are correct because the global inertial frame coordinate system used internal to ElastoDyn (ED) does not match the global inertial frame coordinate system used by OpenFAST (OF).  Basically, X_ED = X_OF, Y_ED = Z_OF, and Z_ED = -Y_OF.  So, you'll notice in the calculation of `TransMat` for platform orientation pitch (P) and yaw (Y) are switched with the sign flipped on P.

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

Thank you so much for the explanation I appreciate your time and responses. I am quite sure the code is correct I am just trying to understand the transformation to be able to get the same loads at the sea water level. 

My main question is why is the sign of the platform pitch and the sign of the ThetaFA different?
If the ThetaFA is positive and the platform pitch is positive I thought the rotation is in the same direction clockwise.

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

This is because a positive `ThetaFA` rotation is defined about the positive Z_ED axis (which is about negative -Y_OF).

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

Thanks a lot. I believe this answers my question and explains it all clearly.

I was confused because in my system where I use the OF coordinate system. I can successfully transfer the aerodynamic loads and have the same values as OF at sea water level when I use PtfmPitch as positive and TTDspPtch as negative. I believe since TTDspPtch and ThetaFA are equivalent this makes sense. 

Best regards,
